etna/packages/etna-js/utils/cancellable_helpers.jsx

```diff
@@ -1,16 +1,12 @@
 import 'regenerator-runtime';
 import {Cancellable, cancelledAsMaybe} from "./cancellable";
-import {useState, useEffect, useCallback} from 'react';
+import {useState, useEffect, useCallback, useRef} from 'react';
 
 export function useAsyncCallback(fn, deps, cleanup = () => null) {
-  const [_, setContext] = useState(() => new Cancellable());
+  const context = useRef(new Cancellable());
   const cancel = useCallback(() => {
-    const newCancellable = new Cancellable();
-    setContext(running => {
```

Comment:
Here's the bug.

Normally, react state hooks provide callbacks, "setters", that allow you to update the state value while /also/ fetching the current value.  Here I used that trick to "fetch" the current cancellable while returning a new cancellable, executing the cancel on the existing value before returning the new one.

but here's the catch -- react state callbacks *do not execute synchronously when you call them!*.  When you run `setX()` react does *NOT* actually set the value immediately or execute the callback immediately.  this blew my mind.

Instead, it simply *buffers the fact that you want to execute it*, and later runs the callback *once* inside of the render loop.

In other words, if you call setX(1) setx(2) and setX(3), react simply accepts the LAST set call setX(3) and runs that during the next render loop, ignoring all the previous set calls.  This optimization normally is convenient for reducing unnecessary renders, but in our case it's problematic because if we happen to cancel twice in the same render loop, we want to cancel twice!  We need to cancel each and every context we create, even if it happens to occur quickly.

That's where `context.cancellable` comes into play.  We can mutate and acccess it synchronously like we expect, react does not deferring or batching, allowing us to consistently make the call we need.